server/pulp/plugins/rsync/configuration.py

```diff
@@ -0,0 +1,165 @@
+import logging
+import os
+
+_LOG = logging.getLogger(__name__)
+
+
+class OneOfValidation(object):
+    """
+    Validates that the the value is one of possible values
+    """
+    def __init__(self, values):
+        self.values = values
+
+    def __call__(self, value, config):
+        if value in self.values:
+            return (True, None)
+        else:
+            return (False, self._err(value))
+
+    def _err(self, value):
+        values_str = ", ".join(self.values)
+        return "{value} is not in allowed values: {allowed}" % (value, values_str)
+
+
+class NonEmptyValidation(object):
+    """
+    Validates that the value is not None
+    """
+    def __call__(self, value, config):
+        if value:
+            return (True, None)
+        else:
+            return (False, self._err(value))
+
+    def _err(self, value):
+        return "attribute cannot be empty"
+
+
+class TypeValidation(object):
+    """
+    Validates that the value is one of allowed types
+    """
+    def __init__(self, allowed_types):
+        self.allowed_types = allowed_types
+
+    def __call__(self, value, config):
+        if [True for _type in self.allowed_types if issubclass(type(value), _type)]:
+            return (True, None)
+        else:
+            return (False, self._err(value))
+
+    def _err(self, value):
+        str_types = ", ".join(self.allowed_types)
+        return "type of atribute(%s) is not in allowed types: %s" % (type(value), str_types)
+
+
+class RequireOptionalIf(object):
+    """
+    Validates that if a particular config is present, other configs that are needed for it are also
+    present.
+    """
+    def __init__(self, required_for_attr, condition):
+        self.required_for_attr = required_for_attr
+        self.condition = condition
+
+    def __call__(self, value, config):
```

Comment:
done. but this should probably be rewritten cause describing the data structure is difficult.